google/cloud/kms_key_name.h

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_KMS_KEY_NAME_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_KMS_KEY_NAME_H
+
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include <iosfwd>
+#include <string>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+
+/**
+ * This class identifies a Google Cloud KMS Key
+ *
+ * A KMS key is identified by its `project_id`, `location`, `key_ring`,
+ * and `kms_key_name`.
+ *
+ * @note this class makes no effort to validate the components of the key,
+ *     It is the application's responsibility to provide a valid project id,
+ *     location, key ring, and KMS key name. Passing invalid values will not
+ *     be checked until the key is used in a RPC.
+ *
+ * See https://cloud.google.com/kms/docs for more information on KMS.
+ */
+class KmsKeyName {
+ public:
+  /**
+   * Constructs a KmsKeyName object identified by the given @p project_id,
+   * @p location, @p key_ring, and @p kms_key_name.
+   */
+  KmsKeyName(std::string const& project_id, std::string const& location,
+             std::string const& key_ring, std::string const& kms_key_name);
+
+  /**
+   * Constructs a KmsKeyName from the given @p full_name.
+   * Returns a non-OK Status if `full_name` is improperly formed.
+   */
+  static StatusOr<KmsKeyName> FromString(std::string full_name);
```

Comment:
Nit: Stylistically, we rarely use class-static factories. We usually have them as non-members[1]. This would fit in better if it were a non-member `StatusOr<KmsKeyname> MakeKmsKeyName(std::string)`

[1]: go/api-design-tips (sorry, internal doc) in the section titled "Prefer non-member factory functions"